src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -296,35 +396,26 @@ object Task {
       * @param taskId The ID of the previous task that was used to match offers.
       */
     def forResidentTask(taskId: Task.Id): Task.Id = {
-      taskId.idString match {
-        case ResidentTaskIdWithInstanceIdRegex(runSpecId, prefix, uuid, container, attempt) =>
-          val newAttempt = attempt.toLong + 1
-          val newIdString = s"$runSpecId.$prefix$uuid.$container.$newAttempt"
-          Task.Id(newIdString)
-        case ResidentTaskIdRegex(runSpecId, separator1, uuid, separator2, attempt) =>
-          val newAttempt = attempt.toLong + 1
-          val newIdString = s"$runSpecId$separator1$uuid$separator2$newAttempt"
-          Task.Id(newIdString)
-        // this is the fallback case and must come second to prevent changing the existing regex
-        case LegacyTaskIdRegex(runSpecId, uuid) =>
-          // the initial dummy task created when reserving does not have an attempt count
-          // when actually launching a task based on a reservation, we consider this attempt #1
-          val newIdString = taskId.idString + ResidentTaskIdAttemptSeparator + "1"
-          Task.Id(newIdString)
-        case _ =>
-          // TODO we cannot really handle this but throwing an exception here is no good.
-          throw new RuntimeException(s"$taskId is no valid resident taskId")
+      taskId match {
+        case EphemeralOrReservedTaskId(instanceId, containerName) =>
+          ResidentTaskId(instanceId, containerName, 1L)
+        case ResidentTaskId(instanceId, containerName, attempt) =>
+          val newAttempt = attempt + 1L
+          ResidentTaskId(instanceId, containerName, newAttempt)
+        case LegacyResidentId(runSpecId, separator, uuid, attempt) =>
+          val newAttempt = attempt + 1L
+          // TODO(karsten): Should we rather upgrade to ResidentTaskId(instanceId, None, 1L)?
```

Comment:
No, it is better to be safe than sorry. If someone calls `idString` of `ResidentTaskId` instance, and the returned string is different, and gets saved somewhere or gets sent to Mesos for whatever reason, and Mesos does byte-by-byte comparison of task labels...

Reply:
✅

Reply:
✅